README.md
# georef-ar-address
[![Build Status](https://travis-ci.org/datosgobar/georef-ar-address.svg?branch=master)](https://travis-ci.org/datosgobar/georef-ar-address)
[![Coverage Status](https://coveralls.io/repos/github/datosgobar/georef-ar-address/badge.svg?branch=master)](https://coveralls.io/github/datosgobar/georef-ar-address?branch=master)
![](https://img.shields.io/github/license/datosgobar/georef-ar-address.svg)
![https://pypi.org/project/georef-ar-address/](https://img.shields.io/pypi/v/georef-ar-address.svg)
![](https://img.shields.io/badge/python-3-blue.svg)

Librería escrita en Python 3 para la identificación de componentes de direcciones argentinas.

## Objetivo

El objetivo de la librería `georef-ar-address` es, dada una dirección en forma de string como entrada, extraer todas las componentes que la componen. **Todo el procesamiento se hace de forma local, sin consultar recursos externos.** Con la librería se intenta cubrir la mayor cantidad de tipos de direcciones posibles, teniendo en cuenta las distintas estructuras que pueden tomar las mismas, así también como errores comunes de escritura, presencia de datos innecesarios y ambigüedades.

Como ejemplo, utilizar la librería sobre la dirección:

`Av. Libertador N1331 2ndo A e/25 de Mayo y Bartolomé Mitre`

Resultaría en la siguiente extracción de componentes:
```
street_names:        ["Av. Libertador", "25 de Mayo", "Bartolomé Mitre"]
door_number.value:   "1331"
door_number.unit:    "N"
floor:               "2ndo A"
type:                "between"
```

## Instalación

La librería se encuentra publicada en [`PyPI`](https://pypi.org/project/georef-ar-address/), y puede ser instalada utilizando `pip`:

```bash
$ pip3 install georef-ar-address
```

## Uso Rápido

Para probar ejemplos de extracción de componentes de forma interactiva, utilizar el siguiente comando:
```bash
$ python -m georef_ar_address
```

Una vez ejecutado el comando se puede escribir una dirección y presionar ENTER para obtener las componentes de la misma.

## Documentación

Para utilizar la librería desde Python, se debe instanciar un objeto de tipo `AddressParser` y utilizar su método `parse`, pasando una dirección como argumento:

```python
>>> from georef_ar_address import AddressParser
>>> parser = AddressParser()
>>> parser.parse('Sarmiento N° 1100')
AddressData({
    "street_names": [
        "Sarmiento"
    ],
    "door_number": {
        "value": "1100",
        "unit": "N°"
    },
    "floor": None,
    "type": "simple"
})
```

El valor de retorno de `parse` es una instancia de `AddressData` conteniendo cada componente de la dirección, en caso de una extracción exitosa. Los campos del objeto `AddressData` son los siguientes:

- `.street_names` *(list)*: Lista de nombres de calles contenidos en la dirección (e.g. `Santa Fe`, `Ruta 4`).
- `.door_number_value` *(str)*: Valor de la altura de la dirección (e.g. `132`, `400/401`, `S/N`)
- `.door_number_unit` *(str)*: Unidad de la altura de la dirección (e.g. `N°`, `nro.`, `Km`).
- `.floor` *(str)*: Piso de la dirección (e.g. `2ndo B`, `PB`).
- `.type` *(str)*: Tipo de dirección detectado. Los valores posibles son:
  - `simple`: Dirección compuesta de un nombre de calle y una altura opcional.
  - `intersection`: Dirección compuesta de dos nombres de calles en forma de intersección, con altura opcional.
  - `between`: Dirección compuesta de tres nombres de calles, especificando una posición sobre una entre otras dos, con altura opcional.

El inicializador de la clase `AddressParser` acepta un parámetro `cache` de tipo `dict` (o equivalente), que le permite cachear internamente resultados de parseos para acelerar el procesamiento de direcciónes con estructuras similares.

## Precisión

La librería `georef-ar-address` (versión `0.0.5`) fue utilizada sobre varios listados de direcciones para poder estimar su precisión al momento de extraer componentes. A continuación, se explica el origen de cada listado y la fidelidad de los datos devueltos por la libería en cada caso:

### Listado de datos abiertos

El primer listado de direcciones que se construyó utiliza varias fuentes provenientes de distintos portales de datos abiertos del país, incluyendo el [Portal Nacional de Datos Abiertos](https://datos.gob.ar/). Se verificó que cada archivo utilizado posea un solo campo de dirección, sin tener la calle separada de la altura.

El listado construido tiene aproximadamente 91000 direcciones. Se utilizó la librería sobre cada una, y los resultados fueron los siguientes:

 - Direcciones categorizadas (no `None`): 95.1%
 - **Aproximado de categorizaciones correctas: 88%**

Para calcular el aproximado de categorizaciones correctas, se tomaron tres muestras de cien direcciones cada una, y se inspeccionó manualmente el resultado brindado por la librería para comprobar que la extración de datos fue correcta. El porcentaje aproximado de 88% categorizaciones correctas es esperado ya que el listado de direcciones construido es muy irregular y contiene grandes cantidades de direcciones escritas de formas impredecibles (es poco uniforme). Esto se debe a que el listado se construyó a partir de datos provenientes de más de 60 archivos distintos, cada uno potencialmente de una fuente distinta.

### Listado de supermercados

El segundo listado de direcciones probado fue el archivo [carrefour.csv](https://gist.github.com/mgaitan/9677204), que contiene la dirección de 325 sucursales del supermercado Carrefour. Los resultados fueron los siguientes:

 - Direcciones categorizadas (no `None`): 99.3%
 - **Categorizaciones correctas: 97.5%**

En este caso, se categorizaron correctamente el 97.5% de las direcciones. El resultado fue comprobado manualmente.

### Listado de sucursales del Banco Nación

El tercer listado de direcciones probado fue el de las [sucursales del Banco de la Nación Argentina](http://www.agencia.mincyt.gob.ar/upload/listado_de_sucursales_bna_web.xls), con 617 elementos. Los resultados fueron los siguientes:

 - Direcciones categorizadas (no `None`): 100%
 - **Categorizaciones correctas: 97.7%**

En este caso, se categorizaron correctamente el 97.7% de las direcciones. El resultado fue comprobado manualmente.

## Diseño

Para leer sobre las desiciones de diseño y funcionamiento de `georef-ar-address`, ver el archivo [**design.md**](docs/design.md).

Para consultar el historial de versiones de `georef-ar-address`, ver el archivo [**history.md**](docs/history.md).

## Soporte
En caso de que encuentres algún bug, tengas problemas con la instalación, o tengas comentarios de alguna parte de `georef-ar-address`, podés mandarnos un mail a [datos@modernizacion.gob.ar](mailto:datos@modernizacion.gob.ar) o [crear un issue](https://github.com/datosgobar/georef-ar-address/issues/new?title=Encontre-un-bug-en-georef-ar-address).


docs/history.md
# Historial de versiones para `georef-ar-address`

## **0.0.9** - 2019/05/07
- Corrige la detección de nombres de calles cuando la calle se llama "calle s n" o contiene "s n".

## **0.0.8** - 2019/04/09
- Mejora la detección de pisos para incluir valores como "12C", "11A", etc. (dos dígitos).
- Mejora la performance de `AddressParser.parse()` cuando se utiliza cache.

## **0.0.7** - 2019/01/29
- Mejora la detección de pisos para incluir valores como "2C", "4A", etc.
- Agrega "pasaje" como tipo de calle.
- Arregla excepción lanzada durante el uso interactivo.

## **0.0.6** - 2019/01/18
- Mejora la detección de detalles de ubicación al final de direcciones de tipo `simple`.
- Modifica valor de retorno de `AddressParser.parse()`: ahora se devuelve un objeto de tipo `AddressData` conteniendo las componentes extraídas de la dirección. Para obtener un diccionario de valores a partir del objeto `AddressData`, utilizar su método `to_dict()`.

## **0.0.5** - 2019/01/14
- Modifica los nombres de los tipos de direcciones `btwn` a `between` e `isct` a `intersection`.
- Remueve restricciones a caracteres aceptados por el método `parse()`: las direcciones como "sÃnta fe 1000" ahora son válidas. Este cambio facilita el procesamiento de direcciones con caracteres erróneos causados por problemas de codificado.
- Corrige interpretación de casos `intersection` donde la segunda calle comienza con "hy" o "y".

## **0.0.4** - 2019/01/10
- El tipo de las direcciones inválidas ahora es `None` en lugar del string `'none'`.

## **0.0.3** - 2019/01/09
- Permite direcciones de tipo `simple` con detalles de ubicación al final. Las siguientes direcciones ahora son interpretadas correctamente:
	- "Periodista Prieto 123 - Partido Lanús"
	- "Ruta 33 s/n Villa Chacón"
- Corrige interpretación de casos `isct` donde la segunda dirección comienza con "h" y se utiliza "e" para separar los nombres, por ejemplo: "Córdoba e Hipólito Yrigoyen".
- Remueve campo `address` redundante de la respuesta de `AddressParser.parse()`.

## **0.0.2** - 2019/01/07
- Agrega uso interactivo con `python3 -m georef_ar_address`.
- Agrega documentación de uso.

## **0.0.1** - 2019/01/04
- Versión inicial.


tests/test_georef_ar_address.py
import json
import os
import unittest
import logging
from georef_ar_address import AddressParser, ADDRESS_TYPES

logging.basicConfig(format='%(message)s',
                    level=os.environ.get('LOG_LEVEL', 'ERROR'))

ADDRESS_DATA_TEMPLATE = {
    'street_names': [],
    'door_number': {
        'value': None,
        'unit': None
    },
    'floor': None
}


def test_file_path(filename):
    return os.path.join(os.path.dirname(__file__), filename)


class CachedAddresParserTest(unittest.TestCase):
    def test_address_parser_cache(self):
        """Al utilizar un cache, se debería agregar una key por cada
        esctructura de dirección distinta."""
        cache = {}
        parser = AddressParser(cache=cache)
        addresses = [
            'Corrientes 1000',
            'Santa fe 2000',
            'callle 10 123 y Tucumán'
        ]

        for address in addresses:
            parser.parse(address)

        self.assertEqual(len(cache), len(addresses))

    def test_address_parser_cache_same_structures(self):
        """Al utilizar un cache, y si dos direcciones comparten la misma
        estructura, solo se debería agregar una key para las dos."""
        cache = {}
        parser = AddressParser(cache=cache)
        addresses = [
            'Corrientes 1000',
            'Tucumán 2000',
            'Córdoba 3333'
        ]

        for address in addresses:
            parser.parse(address)

        self.assertEqual(len(cache), 1)


class AddressParserTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls._parser = AddressParser()

        filename = cls._test_file  # pylint: disable=no-member
        with open(filename) as f:
            cls._test_cases = json.load(f)

        for test_case in cls._test_cases:
            test_case.pop('_comment', None)

        assert all(
            test_case['type'] in ADDRESS_TYPES or test_case['type'] is None
            for test_case in cls._test_cases
        ) and cls._test_cases

        logging.debug('\n%s: %s test cases', os.path.basename(filename),
                      len(cls._test_cases))

    def assert_cases_for_type(self, address_type):
        """Dado un tipo de dirección, leer todos los casos de ese tipo del
        archivo JSON cargado, y comprobar que el parser retorna exactamente
        el valor que se espera.

        Cada test case del archivo JSON contiene un diccionario que tiene la
        misma estructura que la del valor de retorno de AddressParser.parse().
        Utilizando el método 'assert_address_data', se comprueba que el valor
        de retorno de AddressParser.parse() sea idéntico al que está escrito en
        el test case, utilizando el campo 'address' como entrada.

        Los test cases fueron almacenados en archivos JSON para evitar tener
        grandes cantidades de definiciones de variables dentro del código de
        los tests.

        Args:
            address_type (str): Tipo de dirección (None, 'simple',
                'intersection', 'between').

        """
        test_cases = [
            test_case for test_case in self._test_cases
            if test_case['type'] == address_type
        ]

        for test_case in test_cases:
            for key in ADDRESS_DATA_TEMPLATE:
                if key not in test_case:
                    test_case[key] = ADDRESS_DATA_TEMPLATE[key]

            self.assert_address_data(test_case['address'], test_case)

    def assert_address_data(self, address, data):
        parsed = self._parser.parse(address).to_dict()
        parsed['address'] = address
        self.assertDictEqual(parsed, data)


class MockAddressParserTest(AddressParserTest):
    _test_file = test_file_path('test_cases.json')

    def test_simple_cases(self):
        """Comprobar que los casos de tipo 'simple' son parseados
        correctamente."""
        self.assert_cases_for_type('simple')

    def test_intersection_cases(self):
        """Comprobar que los casos de tipo 'intersection' son parseados
        correctamente."""
        self.assert_cases_for_type('intersection')

    def test_between_cases(self):
        """Comprobar que los casos de tipo 'between' son parseados
        correctamente."""
        self.assert_cases_for_type('between')


class RealAddressParserTest(AddressParserTest):
    _test_file = test_file_path('real_cases.json')

    def test_simple_cases(self):
        """Comprobar que los casos de tipo 'simple' son parseados
        correctamente."""
        self.assert_cases_for_type('simple')

    def test_intersection_cases(self):
        """Comprobar que los casos de tipo 'intersection' son parseados
        correctamente."""
        self.assert_cases_for_type('intersection')

    def test_between_cases(self):
        """Comprobar que los casos de tipo 'between' son parseados
        correctamente."""
        self.assert_cases_for_type('between')


class InvalidAddressesParserTest(unittest.TestCase):
    def setUp(self):
        self.parser = AddressParser()

    def test_empty_address(self):
        """Un string vacío como dirección debería resultar en None."""
        data = self.parser.parse('')
        self.assertIsNone(data)

    def test_ambiguous_address(self):
        """Una dirección ambigua debería resultar en None."""
        data = self.parser.parse('Tucumán y Córdoba y Callao')
        self.assertIsNone(data)


class DoorNumberValueTest(unittest.TestCase):
    def setUp(self):
        self.parser = AddressParser()

    def test_int_door_number_value(self):
        """Una dirección con altura numérica en forma de número entero debería
        retornar tipo int en 'normalized_door_number_value()'."""
        data = self.parser.parse('Callao 1231')
        self.assertEqual(data.normalized_door_number_value(), 1231)

    def test_float_door_number_value_comma(self):
        """Una dirección con altura numérica en forma de número decimal debería
        retornar tipo float en 'normalized_door_number_value()' (con coma)."""
        data = self.parser.parse('Ruta provincial 4 km 32,5')
        self.assertAlmostEqual(data.normalized_door_number_value(), 32.5)

    def test_float_door_number_value_point(self):
        """Una dirección con altura numérica en forma de número decimal debería
        retornar tipo float en 'normalized_door_number_value()' (con punto)."""
        data = self.parser.parse('Ruta provincial 4 km 32.5')
        self.assertAlmostEqual(data.normalized_door_number_value(), 32.5)

    def test_none_door_number_value(self):
        """Una dirección sin valor numérico debería retornar 'None' en
        'normalized_door_number_value()'."""
        data = self.parser.parse('Leandro Alem S/N')
        self.assertIsNone(data.normalized_door_number_value())


tools/benchmark.py
import timeit
from georef_ar_address import AddressParser

ITERATIONS = 500


def benchmark(use_cache):
    parser = AddressParser({} if use_cache else None)

    def test_fn():
        parser.parse('Tucuman')
        parser.parse('Tucuman 1000')
        parser.parse('Tucuman Nro 1000')
        parser.parse('Tucuman Nro 1000 2 c')
        parser.parse('Tucuman y Salta')
        parser.parse('Tucuman y Salta 1000')
        parser.parse('Tucuman 1000 y Salta')
        parser.parse('Tucuman e/ Corrientes y Salta')
        parser.parse('Tucuman e/ Corrientes y Salta 1000')
        parser.parse('Tucuman 1000 e/ Corrientes y Salta')

    result = timeit.timeit(stmt=test_fn, number=ITERATIONS)

    print('- Total:         {0:.7f}'.format(result))
    print('- Por iteración: {0:.7f}'.format(result / ITERATIONS))
    print('- Por dirección: {0:.7f}'.format((result / ITERATIONS) / 10))


if __name__ == '__main__':
    print('Sin Cache:')
    benchmark(False)
    print('Con Cache:')
    benchmark(True)


georef_ar_address/__main__.py
"""Módulo '__main__' de georef-ar-address.

Define una función para probar la librería de forma interactiva.

"""

from .address_parser import AddressParser


def address_parser_repl():
    """Función de prueba para probar ejemplos de direcciones en la consola."""
    import json
    parser = AddressParser(cache={})

    print('Ingresar una dirección y presionar [ENTER] para extraer '
          'sus componentes.')
    while True:
        try:
            address = input('> ')
        except (KeyboardInterrupt, EOFError):
            break

        if not address:
            break

        data = parser.parse(address)
        if data:
            print(json.dumps(data.to_dict(), indent=4, ensure_ascii=False,
                             sort_keys=True))
        else:
            print('Error: dirección inválida.')

    print()


address_parser_repl()


georef_ar_address/address_data.py
"""Módulo address_data.py de georef-ar-address

Contiene la clase 'AddressData', utilizada para almacenar, simplificar y
presentar de forma estandarizada los datos producidos por la clase
'AddressParser'.

"""

import re

ADDRESS_TYPES = ['simple', 'intersection', 'between']


class AddressData:
    """Contiene las componentes de una dirección, luego de ser extraídas
    mediante el método 'parse()' de 'AddressParser'.

    Attributes:
        _type (str): Tipo de la dirección.
        _street_names (list): Lista de nombres de calles de la dirección.
        _door_number_value (str): Valor de la altura de la dirección.
        _door_number_unit (str): Unidad de la altura de la dirección.
        _floor (str): Piso de la dirección.

    """

    __slots__ = [
        '_type',
        '_street_names',
        '_door_number_value',
        '_door_number_unit',
        '_floor'
    ]

    def __init__(self, address_type, street_names=None, door_number=None,
                 floor=None):
        """Inicializa un objeto de tipo 'AddressData'.

        Args:
            address_type (str): Ver atributo '_type'.
            street_names (list): Ver atributo '_street_names'.
            door_number (tuple): Tupla de (str, str), donde el primer elemento
                se asigna a '_door_number_value' y el segundo a
                '_door_number_unit'.
            floor (str): Ver atributo '_floor'.

        Raises:
            ValueError: Si el tipo de dirección no es válido.

        """
        if address_type not in ADDRESS_TYPES:
            raise ValueError('Unknown address type: {}'.format(address_type))

        self._type = address_type
        self._street_names = street_names or []
        self._door_number_value = door_number[0] if door_number else None
        self._door_number_unit = door_number[1] if door_number else None
        self._floor = floor

    def to_dict(self):
        """Devuelve una representación del objeto en forma de diccionario. Útil
        para cuando es necesario serializar el objeto a formatos como JSON u
        otros. La estructura del diccionario es la siguiente:

        Entrada: AddressData de 'Tucumán 1300 1° A'

        Salida:
        {
            "door_number": {
                "unit": null,
                "value": "1300"
            },
            "floor": "1° A",
            "street_names": [
                "Tucumán"
            ],
            "type": "simple"
        }

        Returns:
            dict: Diccionario con los componentes de la direccion.

        """
        return {
            'type': self._type,
            'street_names': self._street_names,
            'door_number': {
                'value': self._door_number_value,
                'unit': self._door_number_unit
            },
            'floor': self._floor
        }

    def normalized_door_number_value(self):
        """Retorna el valor de la altura de la dirección como tipo numérico, si
        es posible. Por ejemplo:

        "1000" -> 1000  (int)
        "1,3"  -> 1.3   (float)
        "43.5" -> 43.5  (float)
        "S/N"  -> None

        Returns:
            int, float, NoneType: Valor numérico de la dirección si es posible
                leerlo, None en caso contrario.

        """
        if not self._door_number_value:
            return None

        # Intentar leer un float:
        match = re.search(r'\d+[,.]\d+', self._door_number_value)
        if match:
            return float(match.group(0).replace(',', '.'))

        # Intentar leer un int:
        match = re.search(r'\d+', self._door_number_value)
        return int(match.group(0)) if match else None

    @property
    def type(self):
        """Getter para el atributo '_type'.

        Returns:
            str: Tipo de la dirección.

        """
        return self._type

    @property
    def street_names(self):
        """Getter para el atributo '_street_names'.

        Returns:
            list: Lista de nombre de calles de la dirección.

        """
        return self._street_names

    @property
    def door_number_value(self):
        """Getter para el atributo '_door_number_value'.

        Returns:
            str: Valor de la altura de la dirección.

        """
        return self._door_number_value

    @property
    def door_number_unit(self):
        """Getter para el atributo '_door_number_unit'.

        Returns:
            str: Unidad de la altura de la dirección.

        """
        return self._door_number_unit

    @property
    def floor(self):
        """Getter para el atributo '_floor'.

        Returns:
            str: Piso de la dirección.

        """
        return self._floor

    def __repr__(self):
        return 'AddressData({})'.format(self.to_dict())


georef_ar_address.py
"""Módulo georef_ar_address.py

Contiene clases y funciones utilizadas para extraer información de direcciones
de calles en Argentina. Los tipos de direcciones aceptadas son:

<calle>
<calle> <altura>
<calle> y <calle>
<calle> <altura> y <calle>
<calle> y <calle> <altura>
<calle> entre <calle> y <calle>
<calle> <altura> entre <calle> y <calle>
<calle> entre <calle> y <calle> <altura>
"""

import re
import os
import copy
import nltk

GRAMMARS_DIR = 'grammars'
GRAMMAR_PATH = os.path.join(os.path.dirname(__file__), GRAMMARS_DIR,
                            'address-ar.cfg')
START_PRODUCTION = 'address'

ADDRESS_DATA_TEMPLATE = {
    'address': None,
    'street_names': [],
    'door_number': {
        'value': None,
        'unit': None
    },
    'floor': None,
    'type': 'none'
}

_SEPARATION_REGEXP = r'([^\W\d]{2,}\.?)(\d)'

_NORMALIZATION_REGEXPS = [
    r'\((ex|antes|frente|mano|(al\s)?lado).+?\)',  # Remover aclaraciones
    r'\([sneo]\)',  # Remover aclaraciones de orientación
    r',(\s|$)|\s,',  # Comas utilizadas para separar texto
    r'[()"?]',  # Caracteres no deseados
    r'-+$',  # Guiones al final
    r'\s-\s',  # Guiones entre espacios
    r'(b[°ºª]|barrio\s|bo\.\s).*',  # Indicadores de barrio
    r'\sal\s+(?=\d)'  # Palabra 'al' antes de un número
]

_TOKEN_TYPES = [
    ('AND_WORD', r'y\s(?=\D)|e\s(?=i)'),
    ('AND_NUM', r'y\s(?=\d)'),
    ('OF', r'de\s'),
    ('FLOOR', r'piso(\s|$)'),
    ('DOOR_TYPE', r'(d(e?p)?to\.?|departamento|oficina|of\.)\s'),
    ('GROUNDL', r'(p\.?b\.?|planta\sbaja)(\s|$)'),
    ('ISCT_SEP', r'esquina|esq\.|esq\s|esq/'),
    ('BTWN_SEP', r'e/(calles)?|entre\scalles'),
    ('BETWEEN', r'entre\s'),
    ('KM', r'kil[oó]metro|km\.?'),
    ('MISSING_NAME', r's/nombre'),
    ('MISSING_NUM', r'(sin\s|s/)(n[uú]mero|n(ro\.?|[°º]))'),
    ('S_N', r's[/-]n|sn(\s|$)'),
    ('STREET_TYPE_S', r'(avda|av|bv|diag)[\s.]'),
    ('STREET_TYPE_L', r'calle\s|avenida|bo?ulevard?|diagonal'),
    ('ROUTE', r'ruta|(rta|rn|rp)[\s.]'),
    ('NUM_LABEL_S', r'n\s?[°ºª*]|#|n(?=\d)'),
    ('NUM_LABEL_L', r'nro[\s.]|n[uú]mero'),
    ('DECIMAL', r'\d+[.,]\d+'),
    ('NUM_RANGE', r'\d+[/-]\d+([/-]\d+)*'),
    ('ORDINAL', r'\d+(era?|nd[oa]|[nmtvr][oa])(\s|$|\.)'),
    ('NUM', r'\d+((\s|$)|[°º])'),
    ('N', r'n\s'),
    ('LETTER', r'[^\d\W](\s|$|\.)'),
    ('WORD', r'(\w|\.|\'|`|´|:|/)+'),
    ('WS', r'\s'),
    ('UNKNOWN', '.+')
]


class InvalidTokenException(Exception):
    pass


class InvalidGrammarException(Exception):
    pass


def with_labels(labels):
    return lambda t: t.label() in set(labels)


def load_grammar(grammar_path):
    grammar = nltk.data.load('file:{}'.format(grammar_path))

    if grammar.start().symbol() != START_PRODUCTION:
        raise InvalidGrammarException('Start rule must be "{}"'.format(
            START_PRODUCTION))

    if not grammar.is_nonempty():
        raise InvalidGrammarException('Empty productions are not allowed')

    nonterminals = set()
    terminals = {token_name for token_name, _ in _TOKEN_TYPES}

    for production in grammar.productions():
        nonterminals.add(production.lhs().symbol())

    for production in grammar.productions():
        for element in production.rhs():
            symbol = str(element)

            if isinstance(element, nltk.Nonterminal):
                if symbol not in nonterminals:
                    raise InvalidGrammarException(
                        'Invalid nonterminal: {}'.format(symbol))
            elif symbol not in terminals:
                raise InvalidGrammarException(
                    'Invalid terminal: {}'.format(symbol))

    return grammar


class TreeVisitor:
    __slots__ = ['_tree', '_rank']

    def __init__(self, tree):
        self._tree = tree
        self._rank = None

    def extract_data(self, tokens):
        tree = self._tree.copy(deep=True)

        # Agregar valores reales a las hojas del árbol
        # TODO: Cachear los índices de las hojas de cada subarbol y usar
        # eso directamente.
        for i, tree_pos in enumerate(tree.treepositions('leaves')):
            tree[tree_pos] = tokens[i][0]

        street_names = []
        door_number_value = None
        door_number_unit = None
        floor = None
        condition = with_labels([
            'street',
            'door_number_value',
            'door_number_unit',
            'floor'
        ])

        for subtree in tree.subtrees(condition):
            label = subtree.label()
            subtree_text = ' '.join(subtree.leaves())

            if label == 'street':
                street_names.append(subtree_text)
            elif label == 'door_number_value':
                door_number_value = subtree_text
            elif label == 'door_number_unit':
                door_number_unit = subtree_text
            elif label == 'floor':
                floor = subtree_text

        return street_names, door_number_value, door_number_unit, floor

    def _get_rank(self):
        has_door_number = False
        unnamed_streets = 0

        condition = with_labels(['street_no_num', 'street_with_num'])
        for subtree in self._tree.subtrees(condition):
            if subtree.label() == 'street_with_num':
                has_door_number = True

            if subtree[0][0].label() == 'unnamed_street':
                unnamed_streets += 1

        if has_door_number:
            ranks = ['isct', 'simple', 'btwn']
        else:
            ranks = ['simple', 'isct', 'btwn']

        rank = ranks.index(self.address_type)

        return (unnamed_streets, int(has_door_number), rank)

    @property
    def rank(self):
        if not self._rank:
            self._rank = self._get_rank()

        return self._rank

    @property
    def address_type(self):
        return self._tree.label()


class AddressParser:
    def __init__(self, cache=None):
        self._parser = nltk.EarleyChartParser(load_grammar(GRAMMAR_PATH))

        self._token_regexp = re.compile(
            '|'.join('(?P<{}>{})'.format(*tt) for tt in _TOKEN_TYPES),
            re.IGNORECASE)

        self._separation_regexp = re.compile(_SEPARATION_REGEXP, re.IGNORECASE)

        self._normalization_regexp = re.compile(
            '|'.join(_NORMALIZATION_REGEXPS),
            re.IGNORECASE
        )

        self._cache = cache

    def _tokenize_address(self, address):
        tokens = []
        for mo in self._token_regexp.finditer(address):
            kind = mo.lastgroup
            value = mo.group().strip()

            if kind == 'UNKNOWN':
                raise InvalidTokenException('Value: {}'.format(value))
            elif kind != 'WS':
                tokens.append((value, kind))

        return tokens

    def _preprocess_address(self, address):
        # Reemplazar partes no deseadas por espacios
        normalized = self._normalization_regexp.sub(' ', address)

        # Separar dos o más letras pegadas a números (en ese orden):
        # Sí: 'hola123' -> 'hola 123'
        # Sí: 'ruta nac.3' -> 'ruta nac. 3'
        # No: '1ro de Mayo' -> '1ro de Mayo'
        # No: 'Lote 14 M2' -> 'Lote 14 M2'
        normalized = self._separation_regexp.sub(r'\1 \2', normalized)

        # Normalizar espacios (también remueve trailing/leading whitespace)
        return ' '.join(normalized.split())

    def _disambiguate_trees(self, visitors):
        if len(visitors) > 1:
            visitors.sort(key=lambda v: v.rank, reverse=True)

            # La lista de árboles ahora está ordenada de mejor a peor. Comparar
            # el rank (puntaje) del primer elemento con el del segundo: si son
            # iguales, entonces hay dos (o más) árboles con el mismo puntaje
            # maximal. Esto quiere decir que todos estos árboles son una
            # solución viable, pero no es posible distinguir cuál de ellos es
            # el más adecuado. Si sucede esto, devolver None.
            if visitors[0].rank == visitors[1].rank:
                return None

        return visitors[0]

    def _tokens_parse_tree(self, token_types):
        visitors = [
            TreeVisitor(tree[0])  # tree['address']
            for tree in
            self._parser.parse(token_types)
        ]

        return self._disambiguate_trees(visitors) if visitors else None

    def _parse_token_types(self, token_types):
        if self._cache is not None:
            tokens_hash = hash(tuple(token_types))

            if tokens_hash in self._cache:
                return self._cache[tokens_hash]

            tree = self._tokens_parse_tree(token_types)
            self._cache[tokens_hash] = tree
            return tree

        return self._tokens_parse_tree(token_types)

    def parse(self, address):
        # Remover espacios al comienzo y al final
        address = address.strip()
        processed = self._preprocess_address(address)

        data = copy.deepcopy(ADDRESS_DATA_TEMPLATE)
        data['address'] = address

        if not processed:
            return data

        try:
            tokens = self._tokenize_address(processed)
        except InvalidTokenException:
            return data

        visitor = self._parse_token_types([
            t_type for _, t_type in tokens
        ])

        if visitor:
            street_names, door_number_value, door_number_unit, floor = \
                visitor.extract_data(tokens)
            data['type'] = visitor.address_type
            data['street_names'] = street_names
            data['door_number']['value'] = door_number_value
            data['door_number']['unit'] = door_number_unit
            data['floor'] = floor

        return data
